[BUG]: Fetch API cannot load, URL scheme "file" is not supported.

### Checked for duplicates?

- [X] This issue is not a duplicate

### Does it also happen in the desktop version?

- [X] This bug does not occur in the latest version of Anki

### What are the steps to reproduce this bug?

I use interactive buttons inside my card, And I have button to access to _collection.html from media folder and grab content from there, It was working perfectly, until I updated Ankidroid to 2.18 (as I think, but I don't remember the exact version, about 2 months ago it was working with the latest version), when I tried to remote debug, it shows this:

```
Fetch API cannot load file:///storage/emulated/0/AnkiDroid/collection.media/_collection.html. URL scheme "file" is not supported.
searchAndShowPopup @ _script.js:93
(anonymous) @ _script.js:62
_script.js:105  Error fetching collection: TypeError: Failed to fetch
    at searchAndShowPopup (_script.js:93:9)
    at HTMLSpanElement.<anonymous> (_script.js:62:11)
```

Is there any solutions ? I need interactive cards.
NOTE: it is working perfectly on Anki desktop 24 (latest version as of today)

### Expected behaviour

be able to access _collection.html file without any errors

### Actual behaviour

fails to access the file.

### Debug info

```text
AnkiDroid Version = 2.18beta2 (dc5348852b15e0f27521acd3c5457595f4676f54)

Backend Version = 0.1.38-anki24.04.1 (24.04.1 ccd9ca1a8309b80bcb50ddc5d99c7ce63440bce9)

Android Version = 13 (SDK 33)

ProductFlavor = full

Manufacturer = Xiaomi

Model = M2101K6G

Hardware = qcom

Webview User Agent = Mozilla/5.0 (Linux; Android 13; M2101K6G Build/TKQ1.221013.002; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/124.0.6367.123 Mobile Safari/537.36

ACRA UUID = b8213169-64ed-473b-b9ac-bf2dafbccea7

FSRS Enabled = true

Crash Reports Enabled = true
```


### (Optional) Anything else you want to share?

_No response_

### Research

- [x] I am reporting a bug specific to AnkiDroid (Android app)
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [X] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

--- a/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt
@@ -993,6 +993,7 @@ abstract class AbstractFlashcardViewer :
     }
 
     protected open fun createWebView(): WebView {
+        val resourceHandler = ViewerResourceHandler(this)
         val webView: WebView = MyWebView(this).apply {
             scrollBarStyle = View.SCROLLBARS_OUTSIDE_OVERLAY
             with(settings) {
@@ -1010,7 +1011,7 @@ abstract class AbstractFlashcardViewer :
             isScrollbarFadingEnabled = true
             // Set transparent color to prevent flashing white when night mode enabled
             setBackgroundColor(Color.argb(1, 0, 0, 0))
-            CardViewerWebClient(this@AbstractFlashcardViewer).apply {
+            CardViewerWebClient(resourceHandler, this@AbstractFlashcardViewer).apply {
                 webViewClient = this
                 this@AbstractFlashcardViewer.webViewClient = this
             }
@@ -2233,6 +2234,7 @@ abstract class AbstractFlashcardViewer :
     }
 
     inner class CardViewerWebClient internal constructor(
+        private val resourceHandler: ViewerResourceHandler,
         private val onPageFinishedCallback: OnPageFinishedCallback? = null
     ) : WebViewClient(), JavascriptEvaluator {
         private var pageFinishedFired = true
@@ -2266,6 +2268,7 @@ abstract class AbstractFlashcardViewer :
             if (url.toString().startsWith("file://")) {
                 url.path?.let { path -> migrationService?.migrateFileImmediately(File(path)) }
             }
+            resourceHandler.shouldInterceptRequest(request)?.let { return it }
             return null
         }
 

--- a/AnkiDroid/src/main/java/com/ichi2/anki/ViewerResourceHandler.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/ViewerResourceHandler.kt
@@ -1,0 +1,54 @@
+/*
+ *  Copyright (c) 2024 Brayan Oliveira <brayandso.dev@gmail.com>
+ *
+ *  This program is free software; you can redistribute it and/or modify it under
+ *  the terms of the GNU General Public License as published by the Free Software
+ *  Foundation; either version 3 of the License, or (at your option) any later
+ *  version.
+ *
+ *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along with
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package com.ichi2.anki
+
+import android.content.Context
+import android.webkit.WebResourceRequest
+import android.webkit.WebResourceResponse
+import androidx.core.net.toFile
+import com.ichi2.utils.AssetHelper.guessMimeType
+import timber.log.Timber
+import java.io.FileInputStream
+
+class ViewerResourceHandler(context: Context) {
+    private val mediaDir = CollectionHelper.getMediaDirectory(context).path
+
+    /**
+     * Loads resources from `collection.media` when requested by JS scripts.
+     *
+     * Differently from common media requests, scripts' requests have an `Origin` header
+     * and are susceptible to CORS policy, so `Access-Control-Allow-Origin` is necessary.
+     */
+    fun shouldInterceptRequest(request: WebResourceRequest): WebResourceResponse? {
+        val url = request.url
+        if (request.method != "GET" || url.scheme != "file" || "Origin" !in request.requestHeaders) {
+            return null
+        }
+        try {
+            val file = url.toFile()
+            if (file.parent != mediaDir || !file.exists()) {
+                return null
+            }
+            val inputStream = FileInputStream(file)
+            return WebResourceResponse(guessMimeType(file.path), null, inputStream).apply {
+                responseHeaders = mapOf("Access-Control-Allow-Origin" to "*")
+            }
+        } catch (e: Exception) {
+            Timber.d("File couldn't be loaded")
+            return null
+        }
+    }
+}

--- a/AnkiDroid/src/main/java/com/ichi2/anki/previewer/CardViewerFragment.kt
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/previewer/CardViewerFragment.kt
@@ -23,6 +23,7 @@ import android.webkit.CookieManager
 import android.webkit.WebChromeClient
 import android.webkit.WebResourceError
 import android.webkit.WebResourceRequest
+import android.webkit.WebResourceResponse
 import android.webkit.WebView
 import android.webkit.WebViewClient
 import android.widget.FrameLayout
@@ -37,6 +38,7 @@ import androidx.lifecycle.flowWithLifecycle
 import androidx.lifecycle.lifecycleScope
 import com.ichi2.anki.CollectionHelper
 import com.ichi2.anki.R
+import com.ichi2.anki.ViewerResourceHandler
 import com.ichi2.anki.dialogs.TtsVoicesDialogFragment
 import com.ichi2.anki.localizedErrorMessage
 import com.ichi2.anki.snackbar.showSnackbar
@@ -121,7 +123,15 @@ abstract class CardViewerFragment(@LayoutRes layout: Int) : Fragment(layout) {
     }
 
     private fun onCreateWebViewClient(savedInstanceState: Bundle?): WebViewClient {
+        val resourceHandler = ViewerResourceHandler(requireContext())
         return object : WebViewClient() {
+            override fun shouldInterceptRequest(
+                view: WebView?,
+                request: WebResourceRequest
+            ): WebResourceResponse? {
+                return resourceHandler.shouldInterceptRequest(request)
+            }
+
             override fun onPageFinished(view: WebView?, url: String?) {
                 viewModel.onPageFinished(isAfterRecreation = savedInstanceState != null)
             }

